### PR TITLE
refactor: Paging for the search results and twig templates

### PR DESCRIPTION
**Pull request by aragon999**
### 1. Why is this change necessary?
Just simplify the current setup to work with pagination.

### 2. What does this change do, exactly?
See changelog.

### 3. Describe each step to reproduce the issue or behaviour.
\-

### 4. Please link to the relevant issues (if any).
\-

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

<a href="https://gitpod.io/#https://github.com/shopware/platform/pull/2390"></a>

### CODE REVIEW

File: changelog/_unreleased/2022-03-14-refactor-pagination-templates-and-paging-of-the-search-result.md
Comment (shyim) [inline changelog/_unreleased/2022-03-14-refactor-pagination-templates-and-paging-of-the-search-result.md]:
```suggestion
# Next major version changes
```

File: src/Storefront/Framework/Page/StorefrontSearchResult.php
Comment (shyim) [inline src/Storefront/Framework/Page/StorefrontSearchResult.php]:
```suggestion
 * @deprecated tag:v6.6.0 - will be removed without replacement use `EntitySearchResult` instead, all methods are now contained in the `EntitySearchResult` and the sorting was not in use any more
```

File: src/Core/Framework/DataAbstractionLayer/Search/EntitySearchResult.php
Comment (shyim) [inline src/Core/Framework/DataAbstractionLayer/Search/EntitySearchResult.php]: Is not covered by tests

File: src/Storefront/Resources/views/storefront/component/pagination.html.twig
Comment (shyim) [inline src/Storefront/Resources/views/storefront/component/pagination.html.twig]:
```suggestion
    {% if feature('V6_6_0_0') %}
```

File: src/Storefront/Resources/views/storefront/component/product/listing.html.twig
Comment (shyim) [inline src/Storefront/Resources/views/storefront/component/product/listing.html.twig]:
```suggestion
                                    {% if feature('v6_6_0_0') %}
```

Comment (shyim) [inline src/Storefront/Resources/views/storefront/component/product/listing.html.twig]: Add there also an annotation that this if needs to be removed

File: src/Storefront/Resources/views/storefront/component/review/review.html.twig
Comment (shyim) [inline src/Storefront/Resources/views/storefront/component/review/review.html.twig]:
```suggestion
                                                            {% if feature('v6_6_0_0') %}
```

File: src/Storefront/Resources/views/storefront/page/account/order-history/index.html.twig
Comment (shyim) [inline src/Storefront/Resources/views/storefront/page/account/order-history/index.html.twig]:
```suggestion
                                            {% if feature('v6_6_0_0') %}
```

Comment (shyim) [reply src/Storefront/Resources/views/storefront/page/account/order-history/index.html.twig]: Annotation

File: src/Storefront/Resources/views/storefront/component/review/review.html.twig
Comment (shyim) [reply src/Storefront/Resources/views/storefront/component/review/review.html.twig]: Annotation

File: src/Storefront/Resources/views/storefront/page/product-detail/review/review.html.twig
Comment (shyim) [inline src/Storefront/Resources/views/storefront/page/product-detail/review/review.html.twig]:
```suggestion
                                                            {% if feature('v6_6_0_0') %}
```

**Review by shyim — CHANGES_REQUESTED** [review]

**Comment by shyim** [comment]
Sorry for the late response. I copied from review comments from our internal Gitlab

File: src/Core/Framework/DataAbstractionLayer/Search/EntitySearchResult.php
Comment (aragon999) [reply src/Core/Framework/DataAbstractionLayer/Search/EntitySearchResult.php]: It should be tested by: https://github.com/shopware/platform/pull/2390/files#diff-61eff07120696cdc628dee5f038aebf1fb58ecd30f80f3b2c0a925100648ad33R106-R113

But I moved the test file to the unit test directory (as it only contained unit tests).

**Comment by aragon999** [comment]
I implemented the requested changes and added a comment on one thing which is not clear to me.

Sad that we now have to carry around the `StorefrontSearchResult` one major version longer :-(

**Comment by dfrancos-hub** [comment]
hi @aragon999, please excuse us for the late response. If you plan to continue on this PR, could you resolve the conflicts? thanks in advance 🙏

**Comment by aragon999** [comment]
@dfrancos-hub yes, I mean I fixed and implemented all requested changes.

This PR has some deprecations for at the moment 6.6. As this PR has not been merged in the 6.5 cycle (why?), should these deprecations now target the 6.7 release? (Note that originally this PR was done for 6.4).

**Comment by mitelg** [comment]
@aragon999 as we are already preparing the 6.6 release, your deprecations need to target 6.7.0.

sorry for the inconvenience

**Comment by aragon999** [comment]
@dfrancos-hub @mitelg I have rebased the PR and adapted it to the changes in the meantime.

Furthermore I am not sure if the implementation of the depreciations, is good in this way, or not. Therefore I left them as additional commit.

File: src/Storefront/Framework/Page/StorefrontSearchResult.php
Comment (mitelg) [inline src/Storefront/Framework/Page/StorefrontSearchResult.php]:
```suggestion
            'v6.7.0.0',
```

File: src/Storefront/Framework/Page/StorefrontSearchResult.php
Comment (mitelg) [inline src/Storefront/Framework/Page/StorefrontSearchResult.php]:
```suggestion
            'v6.7.0.0',
```

File: src/Storefront/Framework/Page/StorefrontSearchResult.php
Comment (mitelg) [inline src/Storefront/Framework/Page/StorefrontSearchResult.php]:
```suggestion
            'v6.7.0.0',
```

File: src/Storefront/Framework/Page/StorefrontSearchResult.php
Comment (mitelg) [inline src/Storefront/Framework/Page/StorefrontSearchResult.php]:
```suggestion
            'v6.7.0.0',
```

File: src/Storefront/Resources/views/storefront/component/pagination.html.twig
Comment (mitelg) [inline src/Storefront/Resources/views/storefront/component/pagination.html.twig]:
```suggestion
    {% if feature('v6.7.0.0') %}
```

File: src/Storefront/Resources/views/storefront/component/product/listing.html.twig
Comment (mitelg) [inline src/Storefront/Resources/views/storefront/component/product/listing.html.twig]:
```suggestion
{% if not feature('v6.7.0.0') %}
```

File: src/Storefront/Resources/views/storefront/component/product/listing.html.twig
Comment (mitelg) [inline src/Storefront/Resources/views/storefront/component/product/listing.html.twig]:
```suggestion
                                    {% if feature('v6.7.0.0') %}
```

File: src/Storefront/Resources/views/storefront/component/product/listing.html.twig
Comment (mitelg) [inline src/Storefront/Resources/views/storefront/component/product/listing.html.twig]:
```suggestion
                        {% if feature('v6.7.0.0') %}
```

File: src/Storefront/Resources/views/storefront/component/review/review.html.twig
Comment (mitelg) [inline src/Storefront/Resources/views/storefront/component/review/review.html.twig]:
```suggestion
                                            {% if not feature('v6.7.0.0') %}
```

File: src/Storefront/Resources/views/storefront/component/review/review.html.twig
Comment (mitelg) [inline src/Storefront/Resources/views/storefront/component/review/review.html.twig]:
```suggestion
                                                            {% if feature('v6.7.0.0') %}
```

File: src/Storefront/Resources/views/storefront/page/account/order-history/index.html.twig
Comment (mitelg) [inline src/Storefront/Resources/views/storefront/page/account/order-history/index.html.twig]:
```suggestion
                            {% if not feature('v6.7.0.0') %}
```

File: src/Storefront/Resources/views/storefront/page/account/order-history/index.html.twig
Comment (mitelg) [inline src/Storefront/Resources/views/storefront/page/account/order-history/index.html.twig]:
```suggestion
                                            {% if feature('v6.7.0.0') %}
```

File: src/Storefront/Resources/views/storefront/page/product-detail/review/review.html.twig
Comment (mitelg) [inline src/Storefront/Resources/views/storefront/page/product-detail/review/review.html.twig]:
```suggestion
                                                            {% if feature('v6.7.0.0') %}
```

**Review by mitelg — CHANGES_REQUESTED** [review]

**Comment by dfrancos-hub** [comment]
Hi @aragon999, could you adjust the changelog description to fix the PHPLint check?https://github.com/shopware/shopware/actions/runs/7170576701/job/19523620544?pr=2390 .
Also, the jest admin test is failing probably due to outdated branch, as it doesnt seem to be related to your changes. 
Thanks in advance 🙏

**Comment by aragon999** [comment]
@dfrancos-hub I fixed the changelog lint.

But the admin jest tests are also failing on trunk, so I think this is unrelated.

File: src/Storefront/Page/Product/Review/ReviewLoaderResult.php
Comment (dfrancos-hub) [inline src/Storefront/Page/Product/Review/ReviewLoaderResult.php:13]: Hi @aragon999, am pasting here comments from the internal core team review:

You can do this, but you shouldn't. The use of the FeatureFlags in this MR are not ideal, as they try to change everything completely on the point of v6.7, this will most likely fail on the 6.7 mayor release because this code only runs in tests for months. The aproach should be made more integrational.

File: phpstan-v67-baseline.neon
Comment (dfrancos-hub) [inline phpstan-v67-baseline.neon:4]: Hi @aragon999, am pasting here comments from the internal core team review:

We should adding those triggerDeprecationOrThrow methods.

**Review by dfrancos-hub — CHANGES_REQUESTED** [review]

File: src/Storefront/Page/Product/Review/ReviewLoaderResult.php
Comment (aragon999) [reply src/Storefront/Page/Product/Review/ReviewLoaderResult.php:13]: Mh, but how should the change of the type `StorefrontSearchResult` -> `EntitySearchResult` be implemented otherwise?

Comment (ssltg) [reply src/Storefront/Page/Product/Review/ReviewLoaderResult.php:13]: The `StorefrontSearchResult` already extends from `EntitySearchResult`, so there is no need for a hard rewrite on this step. The use of `StorefrontSearchResult` methods in 6.7 is already captured by the triggerDeprecation methods. So I would just leave the `ReviewLoaderResult` techically as it is and just add a deprecation notice, that it will be extended directly from `EntitySearchResult` from 6.7 on,

```suggestion
* deprecated tag:v6.7.0 - will be extended directly from `EntitySearchResult`.
class ReviewLoaderResult extends StorefrontSearchResult /* extends EntitySearchResult */
{
```

Comment (aragon999) [reply src/Storefront/Page/Product/Review/ReviewLoaderResult.php:13]: @ssltg thank you for the feedback. But in general I meant more the part of the return types:
```php
    public function getReviews(): EntitySearchResult
    {
        return $this->reviews;
    }
```
vs
```php
    public function getReviews(): StorefrontSearchResult
    {
        return $this->reviews;
    }
```

Should I also do it with comments there? But than I do not know how to implement it in the rest of the code?

Comment (ssltg) [reply src/Storefront/Page/Product/Review/ReviewLoaderResult.php:13]: @aragon999 ok, I see your problem here. The reviews are a bit odd implemented here in the first place. Leave the code like it is. Sorry for the inconveniences.

Comment (aragon999) [reply src/Storefront/Page/Product/Review/ReviewLoaderResult.php:13]: @ssltg okay very good :-) Yes it is also implemented the same way in other places, therefore I did not find another solution which was non-breaking.

Regarding the reviews, yes, I once tried to clean that a bit up, but it was also quite challenging: https://github.com/shopware/shopware/pull/2389/files

Maybe once this is merged, I can give this another try in the Shopware 6.6 cycle.

File: phpstan-v67-baseline.neon
Comment (aragon999) [reply phpstan-v67-baseline.neon:4]: @dfrancos-hub @ssltg does it really make sense to throw deprecation errors here? :thinking:

Comment (ssltg) [reply phpstan-v67-baseline.neon:4]: Yes, because we want to keep the phpstan exceptions as low as possible. I know in the current state it is technically not possible that the deprecation is triggered in these methods, but to trigger the deprecation errors helps us to make sure this will stay until 6.7. As many developers are working on our codebase, this is the best way to ensure compatibility even through automatic merge conflict resolves.

**Comment by dfrancos-hub** [comment]
hi @aragon999 , the changes are merged 👍 . Thanks once again for the contribution 🙏

**Pull request closed without merging**
